rust/src/records.rs

```diff
@@ -1,507 +1,537 @@
-    pub profile: CowStr<'a>,
+    while (reader.stream_position()? - pos) < byte_len as u64 {
+        parsed.push(T::read_options(reader, ro, args)?);
+    }
+
+    Ok(parsed)
 }
 
-impl<'a> TryFrom<&'a [u8]> for Header<'a> {
-    type Error = ParseError;
-    fn try_from(value: &'a [u8]) -> Result<Self, Self::Error> {
-        let mut p = Parser::new(value);
-        Ok(Self {
-            library: p.get()?,
-            profile: p.get()?,
-        })
+#[allow(clippy::ptr_arg)] // needed to match binrw macros
+fn write_vec<W: binrw::io::Write + binrw::io::Seek, T: binrw::BinWrite<Args = ()>>(
+    v: &Vec<T>,
+    w: &mut W,
+    opts: &WriteOptions,
+    args: (),
+) -> BinResult<()> {
+    use std::io::SeekFrom;
+
+    let start = w.stream_position()?;
+    (!0u32).write_options(w, opts, args)?; // Revisit...
+    for e in v.iter() {
+        e.write_options(w, opts, args)?;
     }
+    let end = w.stream_position()?;
+    let data_len = end - start - 4;
+    w.seek(SeekFrom::Start(start))?;
+    (data_len as u32).write_options(w, opts, args)?;
+    assert_eq!(w.seek(SeekFrom::End(0))?, end);
+    Ok(())
 }
 
-#[derive(PartialEq, Eq, Debug, Serialize, Deserialize, Clone, IntoStatic)]
+#[derive(Debug, Clone, Eq, PartialEq, BinRead, BinWrite, Serialize, Deserialize)]
+pub struct Header {
+    #[br(map = |s: McapString| s.inner )]
+    #[bw(write_with = write_string)]
+    pub profile: String,
+
+    #[br(map = |s: McapString| s.inner )]
+    #[bw(write_with = write_string)]
+    pub library: String,
+}
+
+#[derive(Debug, Default, Clone, Copy, Eq, PartialEq, BinRead, BinWrite, Serialize, Deserialize)]
 pub struct Footer {
     pub summary_start: u64,
     pub summary_offset_start: u64,
     pub summary_crc: u32,
 }
-impl TryFrom<&[u8]> for Footer {
-    type Error = ParseError;
-    fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
-        let mut p = Parser::new(value);
-        Ok(Self {
-            summary_start: p.get()?,
-            summary_offset_start: p.get()?,
-            summary_crc: p.get()?,
-        })
-    }
-}
 
-#[derive(PartialEq, Eq, Debug, Serialize, Deserialize, Clone, IntoStatic)]
-pub struct Schema<'a> {
+#[derive(Debug, Clone, Eq, PartialEq, BinRead, BinWrite, Serialize, Deserialize)]
+pub struct SchemaHeader {
     pub id: u16,
-    pub name: CowStr<'a>,
-    pub encoding: CowStr<'a>,
-    pub data: Cow<'a, [u8]>,
-}
 
-impl<'a> TryFrom<&'a [u8]> for Schema<'a> {
-    type Error = ParseError;
-    fn try_from(value: &'a [u8]) -> Result<Self, Self::Error> {
-        let mut p = Parser::new(value);
-        Ok(Self {
-            id: p.get()?,
-            name: p.get()?,
-            encoding: p.get()?,
-            data: p.get_byte_array()?,
-        })
+    #[br(map = |s: McapString| s.inner )]
+    #[bw(write_with = write_string)]
+    pub name: String,
+
+    #[br(map = |s: McapString| s.inner )]
+    #[bw(write_with = write_string)]
+    pub encoding: String,
+
+    #[serde(skip)]
+    pub data_len: u32,
```

Comment:
why keep `data_len` around when it's observable from the length of the slice itself?

Reply:
The original idea was to have the header fields match the file spec so that `binrw` could handle all the deserialization. But we need to do some length finagling to properly handle padding anyways. I'll remove it.